File: scriptengine/Ignite/Math/AABB.cs
// Copyright (c) 2026 Evangelion Manuhutu

namespace Ignite;

public struct AABB
{
    public Mathf.Vector3 Min;
    public Mathf.Vector3 Max;

    public AABB(Mathf.Vector3 min, Mathf.Vector3 max)
    {
        Min = min;
        Max = max;
    }

    public Mathf.Vector3 Center => (Min + Max) * 0.5f;
    public Mathf.Vector3 Size => Max - Min;

    public bool IntersectRay(Ray ray, out float distance)
    {
        distance = 0.0f;
        Mathf.Vector3 invDir = new Mathf.Vector3(1.0f / ray.Direction.X, 1.0f / ray.Direction.Y, 1.0f / ray.Direction.Z);
        Mathf.Vector3 tMin = (Min - ray.Origin) * invDir;
        Mathf.Vector3 tMax = (Max - ray.Origin) * invDir;
        Mathf.Vector3 t1 = Mathf.Vector3.Min(tMin, tMax);
        Mathf.Vector3 t2 = Mathf.Vector3.Max(tMin, tMax);
        float tNear = Mathf.Max(Mathf.Max(t1.X, t1.Y), t1.Z);
        float tFar = Mathf.Min(Mathf.Min(t2.X, t2.Y), t2.Z);
        
        if (tNear > tFar || tFar <= 0)
            return false;
        
        distance = tNear;
        return true;
    }
}

File: ignite/engine/src/ignite/math/aabb.cpp
// Copyright (c) 2026 Evangelion Manuhutu

#include "ignite_pch.hpp"
#include "aabb.hpp"

namespace ignite
{
    bool AABB::IntersectRay(const glm::vec3 &rayOrigin, const glm::vec3 &rayDirection, float &outT) const
    {
        // Slab method
        const glm::vec3 invDir = 1.0f / rayDirection;
        glm::vec3 t0s = (min - rayOrigin) * invDir;
        glm::vec3 t1s = (max - rayOrigin) * invDir;

        glm::vec3 tsmaller = glm::min(t0s, t1s);
        glm::vec3 tbigger = glm::max(t0s, t1s);

        float tmin = glm::max(glm::max(tsmaller.x, tsmaller.y), tsmaller.z);
        float tmax = glm::min(glm::min(tbigger.x, tbigger.y), tbigger.z);

        if (tmax >= glm::max(tmin, 0.0f))
        {
            outT = tmin < 0.0f ? tmax : tmin;
            return true;
        }

        return false;
    }

    AABB AABB::Transform(const glm::mat4 &M) const
    {
        glm::vec3 corners[8] = {
            glm::vec3(min.x, min.y, min.z),
            glm::vec3(min.x, min.y, max.z),
            glm::vec3(min.x, max.y, min.z),
            glm::vec3(min.x, max.y, max.z),
            glm::vec3(max.x, min.y, min.z),
            glm::vec3(max.x, min.y, max.z),
            glm::vec3(max.x, max.y, min.z),
            glm::vec3(max.x, max.y, max.z)
        };

        AABB result;
        result.min = glm::vec3(M * glm::vec4(corners[0], 1.0f));
        result.max = result.min;

        for (int i = 1; i < 8; ++i)
        {
            glm::vec3 transformed = glm::vec3(M * glm::vec4(corners[i], 1.0f));
            result.min = glm::min(result.min, transformed);
            result.max = glm::max(result.max, transformed);
        }
        return result;
    }
}
